docs/technical/the-great-stores-migration.md

```diff
@@ -0,0 +1,160 @@
+    components, but it doesn't fit the typical pattern of responding to
+    dispatcher events
+    - `StatsStore` -> `StatsReporter` - this component does not touch
+    application state, and it's actions can be considered side effects.
+
+### Stage 2 - Review and organize `IAppState` shape
+
+The current shape of `IAppState` is relatively flat, and it'd be interesting to
+see if we can organize some of the existing fields into more meaningful objects
+that reflect the structure of the application. This should help us identify
+actions and reducers that perform similar things if we continue down the `redux`
+path.
+
+Some ideas:
+
+ - a `preferences` hash containing settings that the user controls
+
+```ts
+{
+  'preferences': {
+    readonly selectedShell: Shell,
+    readonly selectedExternalEditor?: ExternalEditor,
+    readonly selectedTheme: ApplicationTheme,
+    readonly askForConfirmationOnRepositoryRemoval: boolean,
+    readonly askForConfirmationOnDiscardChanges: boolean
+  }
+}
+```
+
+ - a `userInterface` has representing the current UI state
+
+```ts
+{
+  'userInterface': {
+    readonly windowState: WindowState
+    readonly windowZoomFactor: number
+    readonly appIsFocused: boolean
+    readonly showWelcomeFlow: boolean
+    readonly currentPopup: Popup | null
+    readonly currentFoldout: Foldout | null
+    readonly appMenuState: ReadonlyArray<IMenu>
+    readonly sidebarWidth: number
+    readonly commitSummaryWidth: number
+    readonly titleBarStyle: 'light' | 'dark'
+    readonly isUpdateAvailableBannerVisible: boolean
+  }
+}
+```
+
+### Stage 3 - Extract Functions from `AppStore`
+
+`AppStore` currently rocks in at 4k lines of code, with 180 methods - ~60 of
+which are private. It'd be great to figure out if there's functions in here that
+we can make stateless and extract out of `AppStore`.
+
+This has a few benefits:
+
+ - **coupling** - by forcing us to extract functions, we identify code that is
+ complex or tightly-coupled and might be buggy (race conditions, incorrect flow,
+ etc)
+ - **testability** - functions are easier to test in isolation when you don't
+ have to setup the `AppStore` into the right state
+ - **readability** - by extracting code into functions with meaningful names,
+ the flow of data in `AppStore` becomes easier to reason about
+
+At this point we're still not into shaping our code to fit the `redux` pattern,
+but decomposing `AppStore` to extract functions might help us uncover candidates
+for the next stage.
+
+### Stage 4 - Introduce `redux` to handle some dispatcher work
+
+This is where things get interesting. So rather than trying to rewrite the
+`AppStore` and all the functionality it touches, what if we can setup this flow
+within the app:
+
+ - create a `redux` store alongside `AppStore` that has the same `IAppState`
+   shape
+ - identify a part of `IAppState` that we can port to the Redux pattern
+ - implement actions and reducers to replace this functionality
```

Comment:
You might find [`immer`](https://github.com/mweststrate/immer) useful here.

Reply:
Interesting footnote - `immer` has some confusion about `ReadonlyArray<T>` because the TS declarations assigns every array-like object it finds to `DraftArray<T>`.

<img width="542" src="https://user-images.githubusercontent.com/359239/44289894-93760680-a24c-11e8-8d4f-f104b2271571.png">

This hurts on whenever you want to update an object that has `ReadonlyArray<T>` because it requires manually managing stuff. I'm going to try and upstream a fix for making this handle both `Array<T>` and `ReadonlyArray<T>`.

Reply:
Yeah, that’s unfortunate. Hopefully the issue will be fixable!

Reply:
エラーコードの修正は大変苦労しました（ｐｈｐで）

Reply:
I'm figuring this out here if y'all want to follow along: https://github.com/mweststrate/immer/pull/176